File: src/lib/actions.ts
'use server';

import {
  catalogProduct,
  CatalogProductInput,
  CatalogProductOutput,
} from '@/ai/flows/automated-product-cataloging';
import {
  generateProcessDocumentation,
  GenerateProcessDocumentationInput,
  GenerateProcessDocumentationOutput,
} from '@/ai/flows/process-documentation';
import {
  generateHeritageStory,
  HeritageStorytellingInput,
  HeritageStorytellingOutput,
} from '@/ai/flows/heritage-storytelling';
import {
  analyzeProductPhoto,
  AnalyzeProductPhotoInput,
  AnalyzeProductPhotoOutput,
} from '@/ai/flows/quality-assessment';
import {
  identifyTechnique,
  IdentifyTechniqueInput,
  IdentifyTechniqueOutput,
} from '@/ai/flows/technique-identification';

type FormState<T> = {
  data: T | null;
  error: string | null;
};

export async function catalogProductAction(
  input: CatalogProductInput
): Promise<FormState<CatalogProductOutput>> {
  try {
    const result = await catalogProduct(input);
    return { data: result, error: null };
  } catch (e) {
    const error = e instanceof Error ? e.message : 'An unexpected error occurred.';
    return { data: null, error };
  }
}

export async function generateProcessAction(
  input: GenerateProcessDocumentationInput
): Promise<FormState<GenerateProcessDocumentationOutput>> {
  try {
    const result = await generateProcessDocumentation(input);
    return { data: result, error: null };
  } catch (e) {
    const error = e instanceof Error ? e.message : 'An unexpected error occurred.';
    return { data: null, error };
  }
}

export async function generateStoryAction(
  input: HeritageStorytellingInput
): Promise<FormState<HeritageStorytellingOutput>> {
  try {
    const result = await generateHeritageStory(input);
    return { data: result, error: null };
  } catch (e) {
    const error = e instanceof Error ? e.message : 'An unexpected error occurred.';
    return { data: null, error };
  }
}

export async function assessQualityAction(
  input: AnalyzeProductPhotoInput
): Promise<FormState<AnalyzeProductPhotoOutput>> {
  try {
    const result = await analyzeProductPhoto(input);
    return { data: result, error: null };
  } catch (e) {
    const error = e instanceof Error ? e.message : 'An unexpected error occurred.';
    return { data: null, error };
  }
}

export async function identifyTechniqueAction(
  input: IdentifyTechniqueInput
): Promise<FormState<IdentifyTechniqueOutput>> {
  try {
    const result = await identifyTechnique(input);
    return { data: result, error: null };
  } catch (e) {
    const error = e instanceof Error ? e.message : 'An unexpected error occurred.';
    return { data: null, error };
  }
}

File: backend/src/lib/actions.py
from dataclasses import dataclass
from typing import Generic, Optional, TypeVar
from ai.flows.automated_product_catalog import catalog_product
from ai.flows.process_documentation import generate_process_documentation
from ai.flows.heritage_storytelling import generate_heritage_story
from ai.flows.quality_assessment import analyze_product_photo
from ai.flows.technique_identification import identify_technique

T = TypeVar("T")

@dataclass
class FormState(Generic[T]):
    data: Optional[T]
    error: Optional[str]


# --- Action wrappers ---

async def catalog_product_action(input_data) -> FormState:
    try:
        result = await catalog_product(input_data)
        return FormState(data=result, error=None)
    except Exception as e:
        error_message = str(e) if isinstance(e, Exception) else "An unexpected error occurred."
        return FormState(data=None, error=error_message)


async def generate_process_action(input_data) -> FormState:
    try:
        result = await generate_process_documentation(input_data)
        return FormState(data=result, error=None)
    except Exception as e:
        error_message = str(e) if isinstance(e, Exception) else "An unexpected error occurred."
        return FormState(data=None, error=error_message)


async def generate_story_action(input_data) -> FormState:
    try:
        result = await generate_heritage_story(input_data)
        return FormState(data=result, error=None)
    except Exception as e:
        error_message = str(e) if isinstance(e, Exception) else "An unexpected error occurred."
        return FormState(data=None, error=error_message)


async def assess_quality_action(input_data) -> FormState:
    try:
        result = await analyze_product_photo(input_data)
        return FormState(data=result, error=None)
    except Exception as e:
        error_message = str(e) if isinstance(e, Exception) else "An unexpected error occurred."
        return FormState(data=None, error=error_message)


async def identify_technique_action(input_data) -> FormState:
    try:
        result = await identify_technique(input_data)
        return FormState(data=result, error=None)
    except Exception as e:
        error_message = str(e) if isinstance(e, Exception) else "An unexpected error occurred."
        return FormState(data=None, error=error_message)
